### PR TITLE
Max coverage

### LINKED ISSUE #55: Add missing unit tests for grocy_api_client.py
Some stuff missing: https://coveralls.io/builds/28232582/source?filename=pygrocy/grocy_api_client.py

### PR DESCRIPTION
**Pull request by BlueBlueBlob**
## Description

--Your description here--

**Related issue (if applicable):** fixes #55  and #56 

## Checklist
  - [x] The code change is tested and works locally.
  - [x] tests pass. **Your PR won't be merged unless tests pass**
  - [x] There is no commented out code in this PR

### CODE REVIEW

**Comment by BlueBlueBlob** [comment]
It is easier to test with a trough instance.
Don't need to mock responses, less code.

File: pygrocy/grocy.py [hunk 1]
```diff
@@ -99,9 +99,7 @@ def note(self) -> str:
         return self._note
         
     @property
-    def product(self) -> Product:
-        if self._product_id is None:
-            self.get_details()
+    def product(self) -> ProductData:
```
Comment (SebRut) [inline pygrocy/grocy.py]: I guess it hasn't been enforced in the past, but I initially planned to separate the server data models (stuff ending in Response or Data) from the public interface data models. With this server side changes to not always result in API changes, making it easier to maintain compabilitity. I have to think about this, maybe you could clarif this change.

**Comment by SebRut** [comment]
In relation to this: I want to maintain a set of tests independent of docker stuff or any actual grocy instance for independence. Currently we pull the latest version of grocy for each run and changes to the demo data would lead to tests failing.

File: pygrocy/grocy.py (hunk 1 above)
Comment (BlueBlueBlob) [reply pygrocy/grocy.py]: The code was wrong actually and didn't do nothing... I saw it during testing in coverage report (HTML report is nice :) )

About API change and data model, I understand what's your point.

Have a look to this API : https://github.com/kiwiz/gkeepapi
It uses node for data modeling and request modeling.

**Comment by BlueBlueBlob** [comment]
It's temporary, sorry for the push.
On next Grocy release, it will run only on released version, using the original docker image.

Sorry again, I guess I'm such impatient :)

**Comment by BlueBlueBlob** [comment]
Well,
+7 %
I think I'm done

**Comment by BlueBlueBlob** [comment]
But clearly, we need to split test_grocy and grocy.py

**Comment by SebRut** [comment]
I just dont want to depend all testing on an external source of truth. atm a new grocy release could make all tests fail just by changing the data used in the demo environment. by using tests with mocked responses we can ensure that pygrocy works according to the specs implemented in the test. So docker is for testing compability with current grocy version but our mocked unit tests verify that changes to code don't break stuff (regression). Together we can verify that pygrocy works as expected AND with the current grocy release.

**Comment by BlueBlueBlob** [comment]
I understand.
My arguments :
 + Why asking for an instance for testing ?
 + Why developing on old model data ?
 + At every commit we will know if we are up to date.

 - we need to maintain/check model data on every release.

Hey, it's a discussion, sorry for my English maybe it's rude.

Great regards

**Comment by SebRut** [comment]
> Why asking for an instance for testing ?

For detecting if the lib needs an update to be compatible again, ideally we wouldn't run the docker stuff for PRs but rather periodically, like once a day or automated at every release to test this.

> Why developing on old model data ?
I'm not sure what you mean with this.

**Comment by BlueBlueBlob** [comment]
Ok

"Why will people/we develop on old data model"  ?

But I understand your point of view. But seriously, I will not do the testing mock work : too much lazy work, sorry.

Maybe with 2 branches : one for mock tests and one for real tests. I don't know, let me know :)

**Comment by BlueBlueBlob** [comment]
I'm trying to find a way to have a variable to switch between mock and instance tests.

**Comment by BlueBlueBlob** [comment]
Found this : 
```
@skipIf(SKIP_REAL, 'Skipping tests that hit the real API server.')
```
And it works. Keep in touch

**Comment by BlueBlueBlob** [comment]
If devs want to test trough a real instance : they have to set SKIP_REAL=False in environment variables

**Comment by BlueBlueBlob** [comment]
What's up ? :)

**Comment by SebRut** [comment]
> What's up ? :)
I'm quite busy writing my bachelors thesis, so probably no time to think about this till april :/

**Comment by BlueBlueBlob** [comment]
Good luck 👍

**Comment by BlueBlueBlob** [comment]
I'm closing this pull request. I will do a new one to fix docker pre-release

**Pull request closed without merging**